src/GraphQL/Definitions.hs

```diff
@@ -0,0 +1,222 @@
+  getUnionTypeObjectTypeDefinitions = (getDefinition @a):(getUnionTypeObjectTypeDefinitions @as)
+
+instance UnionTypeObjectTypeDefinitionList '[] where
+  getUnionTypeObjectTypeDefinitions = []
+
+
+-- Interfaces
+class HasInterfaceDefinitions a where
+  getInterfaceDefinitions :: Interfaces
+
+instance forall a as. (HasInterfaceDefinition a, HasInterfaceDefinitions as) => HasInterfaceDefinitions (a:as) where
+  getInterfaceDefinitions = (getInterfaceDefinition @a):(getInterfaceDefinitions @as)
+
+instance HasInterfaceDefinitions '[] where
+  getInterfaceDefinitions = []
+
+class HasInterfaceDefinition a where
+  getInterfaceDefinition :: InterfaceTypeDefinition
+
+instance forall ks fields. (KnownSymbol ks, HasFieldDefinitions fields) => HasInterfaceDefinition (Interface ks fields) where
+  getInterfaceDefinition =
+    let name = Name (toS (symbolVal (Proxy :: Proxy ks)))
+        in InterfaceTypeDefinition name (NonEmptyList (getFieldDefinitions @fields))
+
+-- Give users some help if they don't terminate Arguments with a Field:
+-- NB the "redundant constraints" warning is a GHC bug: https://ghc.haskell.org/trac/ghc/ticket/11099
+instance forall ks t. GHC.TypeLits.TypeError ('GHC.TypeLits.Text ":> Arguments must end with a Field") =>
+         HasFieldDefinition (Argument ks t) where
+  getFieldDefinition = undefined
+
+instance forall ks is ts. (KnownSymbol ks, HasInterfaceDefinitions is, HasFieldDefinitions ts) => HasAnnotatedType (Object ks is ts) where
+  getAnnotatedType =
+    let obj = getDefinition @(Object ks is ts)
+    in TypeNamed (DefinedType (TypeDefinitionObject obj))
+
+instance forall t ks. (KnownSymbol ks, HasAnnotatedType t) => HasFieldDefinition (Field ks t) where
+  getFieldDefinition =
+    let name = Name (toS (symbolVal (Proxy :: Proxy ks)))
+    in FieldDefinition name [] (getAnnotatedType @t)
+
+
+instance forall ks t b. (KnownSymbol ks, HasAnnotatedInputType t, HasFieldDefinition b) => HasFieldDefinition ((Argument ks t) :> b) where
+  getFieldDefinition =
+    let (FieldDefinition name argDefs at) = getFieldDefinition @b
+        argName = Name (toS (symbolVal (Proxy :: Proxy ks)))
+        arg = ArgumentDefinition argName (getAnnotatedInputType @t) Nothing
+    in (FieldDefinition name (arg:argDefs) at)
+
+
+instance forall ks is fields.
+  (KnownSymbol ks, HasInterfaceDefinitions is, HasFieldDefinitions fields) =>
+  HasObjectDefinition (Object ks is fields) where
+  getDefinition =
+    let name = Name (toS (symbolVal (Proxy :: Proxy ks)))
+    in ObjectTypeDefinition name (getInterfaceDefinitions @is) (NonEmptyList (getFieldDefinitions @fields))
+
+-- Builtin output types (annotated types)
+class HasAnnotatedType a where
+  -- TODO - the fact that we have to return TypeNonNull for normal
+  -- types will amost certainly lead to bugs because people will
+  -- forget this. Maybe we can flip the internal encoding to be
+  -- non-null by default and needing explicit null-encoding (via
+  -- Maybe).
```

Comment:
It'll require flipping the internal representation which will then not match 1:1 to the spec (which is Nullable by default). I'm still not sure what the right answer is here, we can discuss on Monday